components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,302 @@
+// Package wheeledodometry implements an odometery estimate from an encoder wheeled base
+package wheeledodometry
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	geo "github.com/kellydunn/golang-geo"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/components/motor"
+	"go.viam.com/rdk/components/movementsensor"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/spatialmath"
+	rdkutils "go.viam.com/rdk/utils"
+)
+
+var (
+	model     = resource.DefaultModelFamily.WithModel("wheeled-odometry")
+	lastLeft  = 0.0
+	lastRight = 0.0
+)
+
+const defaultTimeInt = 500
+
+// Config is used for converting a wheeledodometry MovementSensor config attributes.
+type Config struct {
+	LeftMotors   []string `json:"left_motors"`
+	RightMotors  []string `json:"right_motors"`
+	Base         string   `json:"base"`
+	TimeInterval float64  `json:"time_interval,omitempty"`
+}
+
+type motorPairs struct {
+	left  motor.Motor
+	right motor.Motor
+}
+
+type odometry struct {
+	resource.Named
+	resource.AlwaysRebuild
+
+	lastLeftPos        float64
+	lastRightPos       float64
+	baseWidth          float64
+	wheelCircumference float64
+	timeInterval       float64
+
+	motors []motorPairs
+
+	angularVelocity spatialmath.AngularVelocity
+	acceleration    r3.Vector
+	linearVelocity  r3.Vector
+	position        r3.Vector
+	orientation     spatialmath.EulerAngles
+
+	cancelFunc              func()
+	activeBackgroundWorkers sync.WaitGroup
+	mu                      sync.Mutex
+	logger                  golog.Logger
+}
+
+func init() {
+	resource.RegisterComponent(
+		movementsensor.API,
+		model,
+		resource.Registration[movementsensor.MovementSensor, *Config]{Constructor: newWheeledOdometry})
+}
+
+// Validate ensures all parts of the config are valid.
+func (cfg *Config) Validate(path string) ([]string, error) {
+	var deps []string
+
+	if cfg.Base == "" {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "base")
+	}
+	deps = append(deps, cfg.Base)
+
+	if len(cfg.LeftMotors) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "left motors")
+	}
+	deps = append(deps, cfg.LeftMotors...)
+
+	if len(cfg.RightMotors) == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "right motors")
+	}
+	deps = append(deps, cfg.RightMotors...)
+
+	return deps, nil
+}
+
+// newWheeledOdometry returns a new wheeled encoder movement sensor defined by the given config.
+func newWheeledOdometry(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+	logger golog.Logger,
+) (movementsensor.MovementSensor, error) {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	timeInt := newConf.TimeInterval
+	if timeInt == 0 {
+		timeInt = defaultTimeInt
+	}
+
+	o := &odometry{
+		Named:        conf.ResourceName().AsNamed(),
+		lastLeftPos:  0,
+		lastRightPos: 0,
+		timeInterval: timeInt,
```

Comment:
I changed `timeInterval` to `timeIntervalMSec` but I think since `timeInt` and `defaultTImeInt` are just being used to set `timeIntervalMSec` I don't think they all need to be changed